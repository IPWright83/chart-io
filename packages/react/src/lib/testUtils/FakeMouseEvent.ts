export interface IFakeMouseEventValues {
    pageX?: number;
    pageY?: number;
    offsetX?: number;
    offsetY?: number;
    x?: number;
    y?: number;
}

/**
 * Create a Fake MouseEvent for testing
 */
export class FakeMouseEvent extends MouseEvent {
    /**
     * Construct a new instance of a FakeMoustEvent
     * @param  type        The type of event, e.g. "click", "mouseenter"
     * @param  values      An object containing values for the x,y co-ordinates for the event
     */
    constructor(type: string, values: IFakeMouseEventValues) {
        const { pageX, pageY, offsetX, offsetY, x, y, ...mouseValues } = values;
        super(type, mouseValues);

        Object.assign(this, {
            offsetX: offsetX || 0,
            offsetY: offsetY || 0,
            pageX: pageX || 0,
            pageY: pageY || 0,
            x: x || 0,
            y: y || 0,
        });
    }
}
