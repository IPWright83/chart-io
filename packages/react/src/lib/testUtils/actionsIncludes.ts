/**
 * Assert that all the expected dispatch calls to the Redux store were seen
 * @param {string[]} dispatchCalls The actual calls
 * @param {string[]} expectedCalls The expected calls
 */
export function actionsIncludes(dispatchCalls: string[], expectedCalls: string[]) {
    expectedCalls.forEach((expected) => {
        expect(dispatchCalls).toContain(expected);
        dispatchCalls.splice(dispatchCalls.indexOf(expected), 1);
    });
}
