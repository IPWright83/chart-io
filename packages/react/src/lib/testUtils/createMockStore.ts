import { defaultChartState, defaultEventState, IStore } from "../store";

/**
 * Creates a mock Redux store for testing with
 * @param  state     The initial state of the store
 * @return           A mock store
 */
export function createMockStore(state: { chart?: any; event?: any }): IStore {
    return {
        getState: () => ({
            chart: {
                ...{ ...defaultChartState, id: "test" },
                ...(state.chart ?? {}),
            },
            event: {
                ...defaultEventState,
                ...(state.event ?? {}),
            },
        }),
        dispatch: jest.fn(),
        subscribe: jest.fn(),
    } as unknown as IStore;
}

/**
 * Creates a mock Redux store for Storybook
 * @param  state     The initial state of the store
 * @return           A mock store
 */
export function createMockStorybookStore(state: { chart?: any; event?: any }) {
    return {
        getState: () => ({
            chart: {
                ...defaultChartState,
                ...(state.chart ?? {}),
            },
            event: {
                ...defaultEventState,
                ...(state.event ?? {}),
            },
        }),
        //eslint-disable-next-line @typescript-eslint/no-empty-function
        dispatch: () => {},
        //eslint-disable-next-line @typescript-eslint/no-empty-function
        subscribe: () => {},
    } as unknown as IStore;
}
