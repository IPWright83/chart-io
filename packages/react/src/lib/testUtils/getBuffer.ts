/**
 * Returns the image buffer for a Canvas
 * @param canvas    The HTML canvas that we're going to extract the buffer from
 * @returns         The image buffer
 */
export function getBuffer(canvas: HTMLCanvasElement) {
    const image = canvas.toDataURL("image/png");
    const data = image.replace(/^data:image\/\w+;base64,/, "");
    // eslint-disable-next-line no-undef
    return Buffer.from(data, "base64");
}
