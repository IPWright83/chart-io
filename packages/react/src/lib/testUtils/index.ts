export * from "./actionsIncludes";
export * from "./FakeMouseEvent";
export * from "./createMockStore";
export * from "./getBuffer";
export * from "./wait";
export * from "./renderChart";
export * from "./testEvents";
