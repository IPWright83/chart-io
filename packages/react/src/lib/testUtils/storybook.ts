import { expect, fireEvent } from "@storybook/test";

import { wait } from "./wait";

interface IMouseCoords {
    clientX: number;
    clientY: number;
}

export interface ITestArgs {
    canvasElement: HTMLElement;
    args: {
        onMouseOver: (event: MouseEvent) => void;
        onMouseOut: (event: MouseEvent) => void;
        onClick: (event: MouseEvent) => void;
    };
}

const STORYBOOK_PLAY_DELAY = 800;

/**
 * Creates an interaction test for an SVG chart
 * @param selector  The CSS selector to use to grab an data point
 * @param event     An object containing the mouse coordinates
 * @returns         The Storybook play function
 */
export const createSVGTest =
    (selector: string, event: IMouseCoords) =>
        /**
         * The actual test
         * @param { canvasElement, args } The storybook play function interface
         */
        async ({ canvasElement, args }: ITestArgs) => {
            const eventArgs = { ...event, bubbles: true };

            // Wait for the chart to finish rendering
            await wait(STORYBOOK_PLAY_DELAY);

            // Grab a point
            const dataPoint = canvasElement.querySelector(selector);
            expect(dataPoint).toBeDefined();

            fireEvent(dataPoint, new MouseEvent("mouseover", eventArgs));
            expect(args.onMouseOver).toHaveBeenCalled();

            fireEvent(dataPoint, new MouseEvent("click", eventArgs));
            expect(args.onClick).toHaveBeenCalled();

            fireEvent(dataPoint, new MouseEvent("mouseout", eventArgs));
            expect(args.onMouseOut).toHaveBeenCalled();

            // Add the tooltip back to verify it
            fireEvent(dataPoint, new MouseEvent("mouseover", eventArgs));
            expect(args.onMouseOver).toHaveBeenCalled();
        };

/**
 * Creats an interaction test for the EventReciever in an SVG chart
 * @param event             An object containing the mouse coordinates
 * @param postAssertions    A function to run extra assertions after the mouse events have fired
 * @returns                 The Storybook play function
 */
export const createEventReceiverTest =
    (event: IMouseCoords, postAssertions?: (canvasElement: HTMLElement) => void) =>
        async ({ canvasElement }: { canvasElement: HTMLElement }) => {
            const eventArgs = { ...event, bubbles: true };

            // Wait for the chart to finish rendering
            await wait(STORYBOOK_PLAY_DELAY);

            // Find the event reciver
            const eventReciver = canvasElement.querySelector(".event-receiver");
            expect(eventReciver).toBeDefined();

            fireEvent(eventReciver, new MouseEvent("mouseover", eventArgs));
            await wait(50);

            if (postAssertions) {
                postAssertions(canvasElement);
            }
        };

/**
 * Creates an interaction test for a Canvas chart
 * @param event     An object containing the mouse coordinates
 * @returns         The Storybook play function
 */
export const createCanvasTest =
    (event: IMouseCoords) =>
        /**
         * The actual test
         * @param { canvasElement, args } The storybook play function interface
         */
        async ({ canvasElement, args }: ITestArgs) => {
            const eventArgs = { ...event, bubbles: true };

            // Wait for the chart to finish rendering
            await wait(STORYBOOK_PLAY_DELAY);

            // Grab the virtual canvas
            const virtualCanvas = canvasElement.querySelector(".virtual-canvas");
            expect(virtualCanvas).toBeDefined();

            fireEvent(virtualCanvas, new MouseEvent("mousemove", eventArgs));
            expect(args.onMouseOver).toHaveBeenCalled();
            await wait(50);

            fireEvent(virtualCanvas, new MouseEvent("click", eventArgs));
            expect(args.onClick).toHaveBeenCalled();
            await wait(50);

            fireEvent(virtualCanvas, new MouseEvent("mousemove", { clientX: 0, clientY: 0, bubbles: true }));
            expect(args.onMouseOut).toHaveBeenCalled();
            await wait(50);

            // Add the tooltip back to verify it
            fireEvent(virtualCanvas, new MouseEvent("mousemove", eventArgs));
        };
