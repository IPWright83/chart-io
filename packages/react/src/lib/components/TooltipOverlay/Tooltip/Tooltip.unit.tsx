import type { IColor } from "@d3-chart/types";
import { Provider } from "react-redux";
import React from "react";
import { render } from "@testing-library/react";

import { createMockStore } from "../../../testUtils";

import { Tooltip } from ".";

describe("Tooltip", () => {
    const store = createMockStore({});

    it("should render correctly", async () => {
        const positionStyle = {
            position: "absolute",
            left: 10,
            right: 20,
            top: 30,
            bottom: 40,
        };

        const items = [
            { name: "a", value: 3, icon: "circle" as const, fill: "blue" as IColor },
            { name: "b", value: 8, icon: "square" as const, fill: "orange" as IColor },
            { name: "c", value: 15, icon: "line" as const, fill: "green" as IColor },
        ];

        const { asFragment } = render(
            <Provider store={store}>
                <Tooltip borderColor="red" items={items} positionStyle={positionStyle} />
            </Provider>
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("should render nothing if there are no items", async () => {
        const { asFragment } = render(
            <Provider store={store}>
                <Tooltip items={[]} />
            </Provider>
        );

        expect(asFragment()).toMatchSnapshot();
    });
});
