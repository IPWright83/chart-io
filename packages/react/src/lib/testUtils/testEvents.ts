// prettier-ignore
import { fireEvent } from "@testing-library/react";

import { FakeMouseEvent, IFakeMouseEventValues } from "./FakeMouseEvent";
import { MOUSE_MOVE_THROTTLE } from "../constants";
import { wait } from "./wait";

/**
 * Simulates a click on an element
 * @param  container             The container to apply a css selector to
 * @param  selector              The css selector to apply to find the element
 * @param  callback              The callback that we expect to have been called
 * @param  expected              The arguments that we expect the callback to have been called with
 * @param  fakeMouseEventData    A fake mouse event data object
 */
export async function testMouseClick(
    container: HTMLElement,
    selector: string,
    callback: () => void,
    expected: unknown,
    fakeMouseEventData?: IFakeMouseEventValues
) {
    const element = container.querySelector(selector);

    if (fakeMouseEventData) {
        fireEvent(element, new FakeMouseEvent("click", fakeMouseEventData));
    } else {
        fireEvent.click(element);
    }

    expect(callback).toHaveBeenCalledWith(expected, expect.anything(), expect.anything());
}

/**
 * Simulates a mouse over on an element
 * @param  container             The container to apply a css selector to
 * @param  selector              The css selector to apply to find the element
 * @param  callback              The callback that we expect to have been called
 * @param  expected              The arguments that we expect the callback to have been called with
 * @param  fakeMouseEventData    A fake mouse event data object
 */
export async function testMouseOver(
    container: HTMLElement,
    selector: string,
    callback: () => void,
    expected: unknown,
    fakeMouseEventData?: IFakeMouseEventValues
) {
    const element = container.querySelector(selector);

    if (fakeMouseEventData) {
        fireEvent(element, new FakeMouseEvent("mousemove", fakeMouseEventData));
    } else {
        fireEvent.mouseOver(element);
    }

    expect(callback).toHaveBeenCalledWith(expected, expect.anything(), expect.anything());
}
/**
 * Simulates a mouse exit on an element
 * @param  container                 The container to apply a css selector to
 * @param  selector                  The css selector to apply to find the element
 * @param  callback                  The callback that we expect to have been called
 * @param  expected                  The arguments that we expect the callback to have been called with
 * @param  fakeMouseEnterEventData   A fake mouse event data object to use during the enter event
 * @param  fakeMouseExitEventData    A fake mouse event data object to use during the exit event
 */
export async function testMouseExit(
    container: HTMLElement,
    selector: string,
    callback: () => void,
    expected: unknown,
    fakeMouseEnterEventData?: IFakeMouseEventValues,
    fakeMouseExitEventData?: IFakeMouseEventValues
) {
    const element = container.querySelector(selector);

    if (fakeMouseEnterEventData) {
        fireEvent(element, new FakeMouseEvent("mousemove", fakeMouseEnterEventData));

        await wait(MOUSE_MOVE_THROTTLE * 2);

        fireEvent(
            element,
            new FakeMouseEvent(
                "mousemove",
                fakeMouseExitEventData ?? {
                    pageX: 95,
                    pageY: 95,
                }
            )
        );
    } else {
        fireEvent.mouseOver(element);
        fireEvent.mouseLeave(element);
    }

    expect(callback).toHaveBeenCalledWith(expected, expect.anything(), expect.anything());
}
