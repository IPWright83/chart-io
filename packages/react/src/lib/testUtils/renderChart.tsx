import type { IData, IScale } from "@chart-io/types";
import { Provider } from "react-redux";
import React from "react";
import { render } from "@testing-library/react";

import { createMockStore } from "./createMockStore";
import { wait } from "./wait";

import { themes } from "../themes";

export async function renderChart({
    children,
    data,
    scales = {},
    store,
}: {
    children: JSX.Element;
    data?: IData;
    scales?: Record<string, IScale>;
    store?: any;
}) {
    const mockStore = createMockStore({
        chart: {
            animationDuration: 0,
            dimensions: {
                height: 200,
                width: 200,
            },
            data,
            scales: Object.entries(scales).reduce((result, [key, scaleFunc]) => {
                result[key] = {
                    domain: scaleFunc.domain(),
                    range: scaleFunc.range(),
                    scale: scaleFunc,
                };

                return result;
            }, {}),
            theme: themes.light,
        },
    });

    const { asFragment, container } = render(
        <Provider store={store ?? mockStore}>
            <svg>{children}</svg>
        </Provider>
    );

    await wait(10);

    return { asFragment, container, store: store ?? mockStore };
}
