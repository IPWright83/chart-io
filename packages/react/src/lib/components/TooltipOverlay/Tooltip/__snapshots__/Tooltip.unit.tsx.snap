// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip should render correctly 1`] = `
<DocumentFragment>
  <div
    class="chart-it tooltip"
    style="border: thin solid red; display: inline-block; position: absolute; left: 10px; right: 20px; top: 30px; bottom: 40px; padding: 8px; max-width: 300px; background: rgb(255, 255, 255); opacity: 0.8;"
  >
    <div
      class="chart-it tooltip-item"
      style="display: flex; flex-direction: row; width: 100%; user-select: none; pointer-events: none; color: rgb(0, 0, 0); font-size: 12px; font-family: sans-serif;"
    >
      <div
        class="chart-it indicator-circle"
        style="background: blue; width: 10px; height: 10px; margin-top: 4px; margin-left: 5px; margin-right: 5px; border-radius: 50%; flex-shrink: 0;"
      />
      <div
        class="chart-it tooltip-values"
        style="display: flex; flex-direction: row; justify-content: space-between; width: 100%; flex-grow: 1; flex-shrink: 1;"
      >
        <span
          class="chart-it tooltip-series-name"
          style="margin-right: 15px; max-width: 215px; white-space: nowrap; overflow: hidden; text-overflow: ellipsis; flex-grow: 1; flex-shrink: 1;"
        >
          a:
        </span>
        <span
          class="chart-it tooltip-series-value"
        >
          3
        </span>
      </div>
    </div>
    <div
      class="chart-it tooltip-item"
      style="display: flex; flex-direction: row; width: 100%; user-select: none; pointer-events: none; color: rgb(0, 0, 0); font-size: 12px; font-family: sans-serif;"
    >
      <div
        class="chart-it indicator-square"
        style="background: orange; height: 10px; width: 10px; margin-top: 4px; margin-right: 5px; margin-left: 5px; flex-shrink: 0;"
      />
      <div
        class="chart-it tooltip-values"
        style="display: flex; flex-direction: row; justify-content: space-between; width: 100%; flex-grow: 1; flex-shrink: 1;"
      >
        <span
          class="chart-it tooltip-series-name"
          style="margin-right: 15px; max-width: 215px; white-space: nowrap; overflow: hidden; text-overflow: ellipsis; flex-grow: 1; flex-shrink: 1;"
        >
          b:
        </span>
        <span
          class="chart-it tooltip-series-value"
        >
          8
        </span>
      </div>
    </div>
    <div
      class="chart-it tooltip-item"
      style="display: flex; flex-direction: row; width: 100%; user-select: none; pointer-events: none; color: rgb(0, 0, 0); font-size: 12px; font-family: sans-serif;"
    >
      <div
        class="chart-it indicator-line"
        style="background: green; width: 10px; height: 3px; margin-top: 7px; margin-left: 5px; margin-right: 5px; flex-shrink: 0;"
      />
      <div
        class="chart-it tooltip-values"
        style="display: flex; flex-direction: row; justify-content: space-between; width: 100%; flex-grow: 1; flex-shrink: 1;"
      >
        <span
          class="chart-it tooltip-series-name"
          style="margin-right: 15px; max-width: 215px; white-space: nowrap; overflow: hidden; text-overflow: ellipsis; flex-grow: 1; flex-shrink: 1;"
        >
          c:
        </span>
        <span
          class="chart-it tooltip-series-value"
        >
          15
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tooltip should render nothing if there are no items 1`] = `<DocumentFragment />`;
